localization/ndt_scan_matcher/src/pose_initialization_module.cpp

```diff
@@ -0,0 +1,140 @@
+
+  if (ndt_ptr_->getInputTarget() == nullptr) {
+    res->success = false;
+    RCLCPP_WARN(logger_, "No InputTarget");
+    return;
+  }
+
+  if (ndt_ptr_->getInputSource() == nullptr) {
+    res->success = false;
+    RCLCPP_WARN(logger_, "No InputSource");
+    return;
+  }
+
+  // mutex Map
+  std::lock_guard<std::mutex> lock(*ndt_ptr_mutex_);
+
+  (*state_ptr_)["state"] = "Aligning";
+  res->pose_with_covariance = align_using_monte_carlo(ndt_ptr_, mapTF_initial_pose_msg);
+  (*state_ptr_)["state"] = "Sleeping";
+  res->success = true;
+  res->pose_with_covariance.pose.covariance = req->pose_with_covariance.pose.covariance;
+}
+
+geometry_msgs::msg::PoseWithCovarianceStamped PoseInitializationModule::align_using_monte_carlo(
+  const std::shared_ptr<NormalDistributionsTransform> & ndt_ptr,
+  const geometry_msgs::msg::PoseWithCovarianceStamped & initial_pose_with_cov)
+{
+  if (ndt_ptr->getInputTarget() == nullptr || ndt_ptr->getInputSource() == nullptr) {
+    RCLCPP_WARN(logger_, "No Map or Sensor PointCloud");
+    return geometry_msgs::msg::PoseWithCovarianceStamped();
+  }
+
+  // generateParticle
+  const auto initial_poses =
+    create_random_pose_array(initial_pose_with_cov, initial_estimate_particles_num_);
+
+  std::vector<Particle> particle_array;
+  auto output_cloud = std::make_shared<pcl::PointCloud<PointSource>>();
+
+  for (unsigned int i = 0; i < initial_poses.size(); i++) {
+    const auto & initial_pose = initial_poses[i];
+    const Eigen::Matrix4f initial_pose_matrix = pose_to_matrix4f(initial_pose);
+    ndt_ptr->align(*output_cloud, initial_pose_matrix);
+    const pclomp::NdtResult ndt_result = ndt_ptr->getResult();
+
+    Particle particle(
+      initial_pose, matrix4f_to_pose(ndt_result.pose), ndt_result.transform_probability,
+      ndt_result.iteration_num);
+    particle_array.push_back(particle);
+    const auto marker_array = make_debug_markers(
+      clock_->now(), map_frame_, tier4_autoware_utils::createMarkerScale(0.3, 0.1, 0.1), particle,
+      i);
+    ndt_monte_carlo_initial_pose_marker_pub_->publish(marker_array);
+
+    auto sensor_points_mapTF_ptr = std::make_shared<pcl::PointCloud<PointSource>>();
+    pcl::transformPointCloud(*ndt_ptr->getInputSource(), *sensor_points_mapTF_ptr, ndt_result.pose);
+    publish_point_cloud(initial_pose_with_cov.header.stamp, map_frame_, sensor_points_mapTF_ptr);
+  }
+
+  auto best_particle_ptr = std::max_element(
+    std::begin(particle_array), std::end(particle_array),
+    [](const Particle & lhs, const Particle & rhs) { return lhs.score < rhs.score; });
+
+  geometry_msgs::msg::PoseWithCovarianceStamped result_pose_with_cov_msg;
+  result_pose_with_cov_msg.header.stamp = initial_pose_with_cov.header.stamp;
+  result_pose_with_cov_msg.header.frame_id = map_frame_;
+  result_pose_with_cov_msg.pose.pose = best_particle_ptr->result_pose;
+  // ndt_pose_with_covariance_pub_->publish(result_pose_with_cov_msg);
```

Comment:
Please remove this comment line. 
`is_local_optimal_solution_oscillation` is not defined.

Reply:
Thanks! Done: https://github.com/autowarefoundation/autoware.universe/pull/2254/commits/2b8359f8054805e39f3e5c9dc63f3fb091fb813f

(Is the comment `is_local_optimal_solution_oscillation is not defined.` is just a mistake?)

Reply:
Nice!
(Yes. Sorry. My comment was mistake. :cry: )